presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketAdapterRecordCursor.java

```diff
@@ -0,0 +1,182 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.RecordCursor;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import io.airlift.slice.Slice;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+
+import java.util.List;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_INVALID_BUCKET_FILES;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveBucketAdapterRecordCursor
+        implements RecordCursor
+{
+    private final RecordCursor delegate;
+    private final int[] bucketColumnIndices;
+    private final List<Class<?>> javaTypeList;
+    private final List<TypeInfo> typeInfoList;
+    private final int tableBucketCount;
+    private final int partitionBucketCount;
+    private final int bucketToKeep;
+
+    private final Object[] scratch;
+
+    public HiveBucketAdapterRecordCursor(
+            int[] bucketColumnIndices,
+            List<HiveType> bucketColumnHiveTypes,
+            int tableBucketCount,
+            int partitionBucketCount,
+            int bucketToKeep,
+            TypeManager typeManager,
+            RecordCursor delegate)
+    {
+        this.bucketColumnIndices = requireNonNull(bucketColumnIndices, "bucketColumnIndices is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        requireNonNull(bucketColumnHiveTypes, "bucketColumnHiveTypes is null");
+        this.javaTypeList = bucketColumnHiveTypes.stream()
+                .map(HiveType::getTypeSignature)
+                .map(typeManager::getType)
+                .map(Type::getJavaType)
+                .collect(toImmutableList());
+        this.typeInfoList = bucketColumnHiveTypes.stream()
+                .map(HiveType::getTypeInfo)
+                .collect(toImmutableList());
+        this.tableBucketCount = tableBucketCount;
+        this.partitionBucketCount = partitionBucketCount;
+        this.bucketToKeep = bucketToKeep;
+
+        this.scratch = new Object[bucketColumnHiveTypes.size()];
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        return delegate.getType(field);
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        while (true) {
```

Comment:
This might need to be interruptable (not yieldable, but interruptable)

Reply:
How do I do that? This function doesn't throw InterruptedException, and it doesn't make sense to return either true or false.

Reply:
Throw a presto exception... if the thread is interrupted the query has been destroyed and we want to stop processing... something like this:
```
if (thread.interrupted()) {
   Thread.currentThread(),interrupt();
   throw new PrestoException(query failed)
}